### ISSUE TITLE
[feature request] Provide functional form of LR scheduler formulas (and reconsider older decisions of not doing it)

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#68332]
I propose to have publicly-facing functions of closed-form scheduling formulas depending only on input iteration number.

This way people who are having troubles with current LR scheduler design (IMO statefulness, problems of overriding learning rates after restoring from checkpoint, inability to handle only specific param groups) can still use the tested formulas and wouldn't need to copy this piece as well.

These formulas are useful also for people doing schedulers for quantities other than LR's.

E.g. that would be `torch.schedulers.polynomial(iteration)` or something like this

**Comment by albanD** [comment]
Hi,

Thanks for the proposal, I agree that this would be quite interesting to have.
Note that we do not have anyone actively working on the learning rate scheduler right now (and they would require quite a bit of work to be re-designed/cleaned up). So this is not something the core team is going to push forward right now but we would be happy to support someone that want to become more involved in the design and maintenance of lr scheduler.

**Comment by vadimkantorov** [comment]
I think the first point should be implementing these "plumbing", stateless API, this will require much less design decisions and will already unblock people experimenting with abstractions without having to reimplement the formulas

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/pull/79817

**Comment by vadimkantorov** [comment]
https://github.com/pytorch/pytorch/issues/79888

**Comment by vadimkantorov** [comment]
Another framework having to re-implement schedulers for predictability, simpler recovery from checkpoints and finer-grained control: https://speechbrain.readthedocs.io/en/latest/_modules/speechbrain/nnet/schedulers.html

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/83159 if pure, closed-form functions for schedules are implemented, they already can be reused for other parameters than lr. Later on, it would make it simpler to experiment on a unifying design for different schedulers

**Comment by vadimkantorov** [comment]
Also, maybe sufficient time has passed to get rid of hacks checking that step is called at the right time (mostly were for backcompat IIUC) and simplify the LR scheduler code drastically, this would have helped to avoid https://github.com/pytorch/pytorch/issues/85410

**Comment by vadimkantorov** [comment]
Found a from-scratch design of param schedulers in facebook's fvcore: https://github.com/facebookresearch/fvcore/blob/main/fvcore/common/param_scheduler.py @albanD

**Comment by vadimkantorov** [comment]
Another reason to popularise simple closed-form scheduling formualas: https://github.com/pytorch/pytorch/issues/102261
